code/game/objects/barricades.dm

```diff
@@ -0,0 +1,379 @@
+	. = ..()
+	if(roundstart_barrier_id)
+		var/list/code_and_frequency_list = GLOB.roundstart_barrier_groups[roundstart_barrier_id]
+		if(!code_and_frequency_list)
+			code_and_frequency_list = list(	rand(1,100), // var/code used by src and /obj/item/device/assembly/signaler
+											rand(1200, 1600)) // var/frequency, just like above
+			GLOB.roundstart_barrier_groups[roundstart_barrier_id] = code_and_frequency_list
+
+		update_wifi_password(code_and_frequency_list[1], code_and_frequency_list[2])
+
+
+/obj/structure/barrier/four_way/Destroy()
+	if(frequency)
+		SSradio.remove_object(src, frequency)
+	. = ..()
+
+
+/obj/structure/barrier/four_way/proc/update_wifi_password(new_code, new_frequency)
+	if(frequency)
+		SSradio.remove_object(src, frequency)
+	code = new_code
+	frequency = new_frequency
+	if(frequency)
+		SSradio.add_object(src, frequency)
+
+
+/obj/structure/barrier/four_way/attackby(obj/item/I, mob/user)
+	if(istype(I, /obj/item/device/assembly/signaler))
+		var/obj/item/device/assembly/signaler/signaler = I
+		if((signaler.code == code) && (signaler.frequency == frequency)) // Linked signaler is used, break the link
+			update_wifi_password()
+			signaler.audible_message("\icon[signaler] *beep*")
+		else if(!frequency || !code) // There is no link, establish one
+			update_wifi_password(signaler.code, signaler.frequency)
+			signaler.audible_message("\icon[signaler] *beep* *beep*")
+		else // Access denied
+			signaler.audible_message("\icon[signaler] *beep* *beep* *beep*") // Silly and cryptic, yet authentic signaler feedback
+		return
+
+	var/list/usable_qualities = list(QUALITY_BOLT_TURNING)
+	if(health < maxHealth)
+		usable_qualities.Add(QUALITY_WELDING)
+	if(code || frequency)
+		usable_qualities.Add(QUALITY_PULSING)
+
+	var/tool_type = I.get_tool_type(user, usable_qualities)
+
+	switch(tool_type)
+		if(QUALITY_WELDING)
+			if(health < maxHealth)
+				if(I.use_tool(user, src, WORKTIME_NORMAL, tool_type, FAILCHANCE_NORMAL,  required_stat = STAT_MEC))
+					user.visible_message(SPAN_NOTICE("\The [user] repairs \the [src]."), SPAN_NOTICE("You repair \the [src]."))
+					health = maxHealth
+					is_damaged = FALSE
+					update_icon()
+					return
+		if(QUALITY_BOLT_TURNING) // Deconstruction and hacking be harder when barrier is deployed
+			if(I.use_tool(user, src, (is_open ? WORKTIME_LONG : WORKTIME_NORMAL), tool_type, (is_open ? FAILCHANCE_CHALLENGING : FAILCHANCE_NORMAL),  required_stat = STAT_MEC))
+				for(var/material_name in matter)
+					var/material/material_datum = get_material_by_name(material_name)
+					material_datum.place_sheet(loc, amount = matter[material_name])
+				user.visible_message(SPAN_NOTICE("\The [user] dismantles \the [src]."), SPAN_NOTICE("You dismantle \the [src]."))
+				qdel(src)
+				return
+		if(QUALITY_PULSING)
+			if(I.use_tool(user, src, (is_open ? WORKTIME_LONG : WORKTIME_NORMAL), tool_type, (is_open ? FAILCHANCE_CHALLENGING : FAILCHANCE_NORMAL),  required_stat = STAT_MEC))
```

Comment:
Maybe use COG here?

Reply:
I did want to use COG initially, but since locker hacking used MEC, I went with it as well, for consistency sake.